File: src/jfan/fan/sys/ZipEntryFile.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   26 Mar 06  Brian Frank  Creation
//
package fan.sys;

/**
 * ZipEntryFile represents a file entry inside a zip file.
 */
public class ZipEntryFile
  extends File
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public ZipEntryFile(java.util.zip.ZipFile parent, java.util.zip.ZipEntry entry)
  {
    super(Uri.fromStr("/" + LocalFile.fileNameToUriName(entry.getName())));
    this.parent = parent;
    this.entry  = entry;
  }

  public ZipEntryFile(Zip parent, java.util.zip.ZipEntry entry)
  {
    super(Uri.fromStr("/" + LocalFile.fileNameToUriName(entry.getName())));
    this.parent = parent;
    this.entry  = entry;
  }

//////////////////////////////////////////////////////////////////////////
// Obj
//////////////////////////////////////////////////////////////////////////

  public Type type() { return Sys.ZipEntryFileType; }

//////////////////////////////////////////////////////////////////////////
// File
//////////////////////////////////////////////////////////////////////////

  public Bool exists()
  {
    return Bool.True;
  }

  public Int size()
  {
    if (isDir().val) return null;
    long size = entry.getSize();
    if (size < 0) return null;
    return Int.pos(size);
  }

  public DateTime modified()
  {
    return DateTime.java(entry.getTime());
  }

  public void modified(DateTime time)
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

  public Str osPath()
  {
    return null;
  }

  public File parent()
  {
    return null;
  }

  public List list()
  {
    return new List(Sys.FileType, 0);
  }

  public File normalize()
  {
    return this;
  }

  public File plus(Uri uri, Bool checkSlash)
  {
    // TODO
    throw UnsupportedErr.make("ZipEntryFile.plus").val;
  }

//////////////////////////////////////////////////////////////////////////
// File Management
//////////////////////////////////////////////////////////////////////////

  public File create()
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

  public File moveTo(File to)
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

  public void delete()
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

  public File deleteOnExit()
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

//////////////////////////////////////////////////////////////////////////
// IO
//////////////////////////////////////////////////////////////////////////

  public Buf open(Str mode)
  {
    throw UnsupportedErr.make("ZipEntryFile.open").val;
  }

  public Buf mmap(Str mode, Int pos, Int size)
  {
    throw UnsupportedErr.make("ZipEntryFile.mmap").val;
  }

  public InStream in(Int bufSize)
  {
    try
    {
      java.io.InputStream in;
      if (parent instanceof Zip)
      {
        // never buffer if using ZipInputStream
        in = new java.io.FilterInputStream(((Zip)parent).zipIn) { public void close() {} };
      }
      else
      {
        in = ((java.util.zip.ZipFile)parent).getInputStream(entry);

        // buffer if specified
        if (bufSize != null && bufSize.val != 0)
          in = new java.io.BufferedInputStream(in, (int)bufSize.val);
      }

      // return as fan stream
      return new SysInStream(in);
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e).val;
    }
  }

  public OutStream out(Bool append, Int bufSize)
  {
    throw IOErr.make("ZipEntryFile is readonly").val;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  final Object parent;
  final java.util.zip.ZipEntry entry;

}

File: src/sys/es/fan/ZipEntryFile.js
//
// Copyright (c) 2023, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   28 Jul 2023  Kiera O'Flynn  Creation
//

/**
 * ZipEntryFile.
 */
class ZipEntryFile extends File {

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  constructor(uri) {
    super(uri);
  }

  #isFileBacked;
  #entry;
  #yauzlZip;
 
  #zip;

  // Entry is parsed from central directory
  static makeFromFile(centralEntry, yauzlZip, zip) {
    const instance = new ZipEntryFile(Uri.fromStr("/" + centralEntry.fileName));
    instance.#isFileBacked = true;
    instance.#entry = centralEntry;
    instance.#yauzlZip = yauzlZip;
    instance.#zip = zip;
    return instance;
  }

  // Entry is parsed from local file header
  static makeFromStream(localEntry, yauzlZip) {
    const instance = new ZipEntryFile(Uri.fromStr("/" + localEntry.fileName));
    instance.#isFileBacked = false;
    instance.#entry = localEntry;
    instance.#yauzlZip = yauzlZip;
    return instance;
  }

//////////////////////////////////////////////////////////////////////////
// Info
//////////////////////////////////////////////////////////////////////////

  exists() {
    return true;
  }

  modified(val) {
    if (val)
      throw IOErr.make("ZipEntryFile is readonly");

    return yauzl.dosDateTimeToFantom(this.#entry.lastModFileDate, this.#entry.lastModFileTime);
  }

  size() {
    if ((this.#entry.generalPurposeBitFlag & 0x8) &&
         !this.#isFileBacked &&
         !this.#entry.foundDataDescriptor)
      return null;
    return this.#entry.uncompressedSize;
  }

  normalize() {
    return this;
  }

  osPath() {
    return null;
  }

//////////////////////////////////////////////////////////////////////////
// Navigation
//////////////////////////////////////////////////////////////////////////

  list(pattern) {
    if (!this.#isFileBacked || !this.uri().isDir())
      return List.make(File.type$, []);

    return this.#zip.contents().findAll((uri, f) => {
      this.uri().equals(uri.parent()) &&
      (!pattern || pattern.matches(uri.name))
    }).vals();
  }

  parent() {
    if (!this.#isFileBacked)
      return null;
    return this.#zip.contents().get(this.uri().parent());
  }

  plus(path, checkSlash=true) {
    if (!this.#isFileBacked)
      return File.make(newUri, checkSlash); // nonexistent file

    const newUri = this.uri().plus(path);
    const a = this.#zip.contents().get(newUri);
    if (a) return a;
    const b = this.#zip.contents().get(newUri.plusSlash());
    if (b) {
      if (checkSlash) throw IOErr.make("Must use trailing slash for dir: " + newUri.toString());
      else return b;
    }
    return File.make(newUri, checkSlash); // nonexistent file
  }

//////////////////////////////////////////////////////////////////////////
// Reading
//////////////////////////////////////////////////////////////////////////

  in$(bufferSize=4096) {
    if (this.#isFileBacked)
      return (this.#in = this.#yauzlZip.getInStream(this.#entry, {}, bufferSize));
    else {
      if (this.#in) throw IOErr.make("In stream already created");
      return (this.#in = this.#yauzlZip.getInStreamFromStream(this.#entry, {}, bufferSize));
    }
  }

  #in;
  __in(bufferSize=4096) {
    return this.#in || this.in$(bufferSize);
  }

//////////////////////////////////////////////////////////////////////////
// Not writing
//////////////////////////////////////////////////////////////////////////

  create() { this.#throwIO("create") }
  delete$() { this.#throwIO("delete") }
  deleteOnExit() { this.#throwIO("deleteOnExit") }
  moveTo() { this.#throwIO("moveTo") }
  out() { this.#throwIO("out") }

  #throwIO(name) { throw IOErr.make(`Cannot call '${name}'; zip entries are readonly.`) }

}